File: 1052-grumpy-bookstore-owner/main.py
class Solution:
    def maxSatisfied(self, customers: List[int], grumpy: List[int], minutes: int) -> int:
        
        n = len(customers)

        satisfied_customers = sum([costumer for idx, costumer in enumerate(customers) if grumpy[idx] == 0])
        additional_satisfied = sum([customers[idx] for idx in range(minutes) if grumpy[idx] == 1])

        max_additional_satisfied = additional_satisfied

        for idx in range(minutes, n):
            if grumpy[idx] == 1:
                additional_satisfied += customers[idx]
            if grumpy[idx - minutes] == 1:
                additional_satisfied -= customers[idx - minutes]

            max_additional_satisfied = max(max_additional_satisfied, additional_satisfied)

        return satisfied_customers + max_additional_satisfied

File: 1052-grumpy-bookstore-owner/main.go
func maxSatisfied(customers []int, grumpy []int, minutes int) int {
    
    var n int = len(customers)

    var satisfiedCustomers int = 0
    for idx, costumer := range customers {
        if grumpy[idx] == 0 {
            satisfiedCustomers += costumer
        }
    }

    var additionalSatisfied int = 0
    for idx := range minutes {
        if grumpy[idx] == 1 {
            additionalSatisfied += customers[idx]
        }
    }

    var maxAdditionalSatisfied int = additionalSatisfied

    for idx := minutes; idx < n; idx++ {
        if grumpy[idx] == 1 {
            additionalSatisfied += customers[idx]
        }

        if grumpy[idx - minutes] == 1 {
            additionalSatisfied -= customers[idx - minutes]
        }

        maxAdditionalSatisfied = max(maxAdditionalSatisfied, additionalSatisfied)

    }

    return maxAdditionalSatisfied + satisfiedCustomers
}

func max(a, b int) int {
    if a < b {
        return b
    }
    return a
}
